### PR TITLE
Create master users model

### PR DESCRIPTION
**Pull request by govindkrjoshi**
Combine User and Admin into a single model
- Adds migrations to create MasterUser, UserAuthentication, PhoneNumberAuthentication (user by mobile app) and EmailAuthentication (for dashboard; devise authentication)
- Adds models and their associations
- Adds tasks to create `MasterUser`s from `User`s and `Admin`s

### CODE REVIEW

File: app/models/phone_number_authentication.rb [hunk 1]
```diff
@@ -0,0 +1,18 @@
+class PhoneNumberAuthentication < ApplicationRecord
+  has_secure_password
+
+  has_one :user_authentication, as: :authenticatable
+  has_one :master_user, through: :user_authentication
+
+  belongs_to :facility, foreign_key: 'registration_facility_id'
+
+  validates :phone_number, presence: true, uniqueness: true, case_sensitive: false
+  validates :password, allow_blank: true, length: { is: 4 }, format: { with: /[0-9]/, message: 'only allows numbers' }
```
Comment (timmyjose) [inline app/models/phone_number_authentication.rb]: Is the password in this case the OTP? If so, currently it is 6 digits long IIRC. Is this length changing to 4 now?

File: app/models/master_user.rb [hunk 2]
```diff
@@ -0,0 +1,35 @@
+class MasterUser < ApplicationRecord
+
+  AUTHENTICATION_TYPES = {
+    emailauthentication: 'EmailAuthentication',
```
Comment (govindkrjoshi) [inline app/models/master_user.rb]:
```suggestion
    email_authentication: 'EmailAuthentication',
```

File: app/models/master_user.rb [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+    phone_number_authentication: 'PhoneAuthentication'
```
Comment (govindkrjoshi) [inline app/models/master_user.rb]:
```suggestion
    phone_number_authentication: 'PhoneNumberAuthentication'
```

File: db/migrate/20190425084108_create_phone_number_authentications.rb [hunk 4]
```diff
@@ -0,0 +1,20 @@
+class CreatePhoneNumberAuthentications < ActiveRecord::Migration[5.1]
+  def change
+    create_table :phone_number_authentications, id: :uuid do |t|
+      t.string :phone_number, unique: true, null: false
+      t.string :password_digest, null: false
+      t.string :otp, null: false
+      t.datetime :otp_valid_until, null: false
+      t.datetime :logged_in_at
+      t.string :access_token, null: false
+
```
Comment (govindkrjoshi) [inline db/migrate/20190425084108_create_phone_number_authentications.rb]:
```suggestion
```

Comment (govindkrjoshi) [inline db/migrate/20190425084108_create_phone_number_authentications.rb]:
```suggestion
```

File: lib/tasks/data_migration.rake [hunk 5]
```diff
@@ -82,6 +82,92 @@ namespace :data_migration do
...
+          :created_at,
+          :updated_at,
+          :deleted_at
+        ))
+
+        phone_number_authentication = PhoneNumberAuthentication.create(user_attributes.slice(
+          :phone_number,
+          :password_digest,
+          :otp,
+          :otp_valid_until,
```
Comment (govindkrjoshi) [inline lib/tasks/data_migration.rake]:
```suggestion
          :otp_valid_until,
          :logged_in_at
```

File: lib/tasks/data_migration.rake [hunk 6]
```diff
@@ -82,6 +82,93 @@ namespace :data_migration do
...
+        master_user.master_user_authentications.create(
+          authenticatable: email_authentication
+        )
+      end
+    end
+  end
+
+  desc "Fix null invited_by for email authentications when migrating from admins"
+  task fix_invited_by_for_email_authentications: :environment do
+    EmailAuthentication.all.take(1).each do |email_authentication|
```
Comment (kitallis) [inline lib/tasks/data_migration.rake]: I think you can just do,
```suggestion
    EmailAuthentication.take(1).each do |email_authentication|
```

**Pull request merged**